jax/experimental/export/export.py

```diff
@@ -989,30 +1004,30 @@ def flattened_primal_fun_jax(*args_flat):
       itertools.chain(in_avals,
                       map(lambda a: a.at_least_vspace(), out_avals)))
 
-  all_in_shardings = expand_in_shardings(in_shardings,
-                                         module_kept_var_idx, len(in_avals))
-  vjp_in_shardings, vjp_out_shardings = canonical_shardings(
-    tuple(itertools.chain(all_in_shardings, out_shardings)),
-    all_in_shardings)
-
   if apply_jit:
+    # Prepare a device assignment. For exporting purposes, all it matters
+    # is the number of devices.
+    device_assignment = jax.devices(jax.default_backend())[:nr_devices]
```

Comment:
I don't think this is correct. `device_assignment` is only always jax.devices(). The devices can be laid out in any order which is usually decided by mesh_utils.

Reply:
Which part is not correct?

I do not understand
> . device_assignment is only always jax.devices()

For lowering purposes it does not matter what is the device assigment, e.g., what devices are in what order, because the HLO contains only logical device ids, from 0 to `len(device_assignment) - 1`. When you compile and run, you need the actual mapping of these logical devices, but not when you lower. This is why I think it is sufficient to work with `nr_devices` and the `HloSharding` protos for export purposes.

The code here is used in some rare circumstances, after you lowered a function this gives you a hook to lower the `vjp`, which will call `export(pjit(vjp(f)))`. For the call to `pjit` we need actual XlaCompatibleSharding, for which we need a device assignment. But we are only going to lower this code now, we won't compile it, or run it, so it does not matter which devices are in the list.

In `export.call_exported` we actually use an `Exported` and there we will use an actual device assignment for the embedding JAX computation, from `ModuleContext.axis_context.device_assignment`.

Reply:
Ohh, sorry I miswrote: I wanted to say: `device_assignment is not always jax.devices()`.

Okay, if this is lowering time, then it's fine. I didn't know this is happening during lowering.

But one thing I would do is test this!

Try with something like: (note that you need to try this on TPU to actually get mesh_utils to return a different order than jax.devices())

```
devices = mesh_utils.create_device_mesh((4, 2))
mesh = Mesh(devices, ('x', 'y')
NamedSharding(mesh, P(None, 'y'))  # use this sharding and check if the HloSharding is the correct one as you expect.
```

Reply:
I shared [a colab](https://colab.corp.google.com/drive/10g9y8YjgV2U4xBzNRb8Qg1AtBD07C9hP?resourcekey=0-j2hhDO-PHfljbxbX4spmmA#scrollTo=uT3Y8Nd-zIG6) showing that a function results in the same exact lowering with two different device assignments.

Reply:
Thanks! Maybe add that as a test too?

Reply:
I don't really think that a test is warranted here, the fact that the StableHLO depends only on the HloSharding and not the actual device assignment is a property to HLO, not something that the export code ensures.